File: 3. Behavioral Design Patterns/Strategy/Kotlin/StrategyExample.kt
// Strategy Pattern Example in Kotlin
// Sorter with BubbleSort and QuickSort strategies

interface SortStrategy { fun sort(data: List<Int>): List<Int> }

class BubbleSort : SortStrategy {
    override fun sort(data: List<Int>): List<Int> {
        val arr = data.toMutableList()
        for (i in arr.indices) {
            for (j in 0 until arr.size - i - 1) {
                if (arr[j] > arr[j + 1]) {
                    val tmp = arr[j]
                    arr[j] = arr[j + 1]
                    arr[j + 1] = tmp
                }
            }
        }
        println("BubbleSort used")
        return arr
    }
}

class QuickSort : SortStrategy {
    override fun sort(data: List<Int>): List<Int> {
        val arr = data.toMutableList()
        arr.sort() // brevity
        println("QuickSort used")
        return arr
    }
}

class Sorter(private var strategy: SortStrategy) {
    fun setStrategy(s: SortStrategy) { strategy = s }
    fun sort(data: List<Int>) = strategy.sort(data)
}

fun main(){
    val data = listOf(5,2,9,1,5,6)
    val sorter = Sorter(BubbleSort())
    println(sorter.sort(data))
    sorter.setStrategy(QuickSort())
    println(sorter.sort(data))
}

File: 3. Behavioral Design Patterns/Strategy/Javascript/StrategyExample.js
// Strategy Pattern Example in JavaScript
// Sorter with BubbleSort and QuickSort strategies

class SortStrategy { sort(data){ return data } }

class BubbleSort extends SortStrategy {
  sort(data){
    const arr = [...data]
    for(let i=0;i<arr.length;i++){
      for(let j=0;j<arr.length-i-1;j++){
        if(arr[j] > arr[j+1]) [arr[j], arr[j+1]] = [arr[j+1], arr[j]]
      }
    }
    console.log('BubbleSort used')
    return arr
  }
}

class QuickSort extends SortStrategy {
  sort(data){
    const arr = [...data].sort((a,b)=>a-b)
    console.log('QuickSort used')
    return arr
  }
}

class Sorter {
  constructor(strategy){ this.strategy = strategy }
  setStrategy(strategy){ this.strategy = strategy }
  sort(data){ return this.strategy.sort(data) }
}

(function main(){
  const data = [5,2,9,1,5,6]
  const sorter = new Sorter(new BubbleSort())
  console.log(sorter.sort(data))
  sorter.setStrategy(new QuickSort())
  console.log(sorter.sort(data))
})()
